app/types/websocket.py

```diff
@@ -0,0 +1,120 @@
+        Broadcast(settings.REDIS_HOST)
+        if settings.REDIS_HOST
+        else Broadcast("memory://")
+    )
+
+    def __init__(self):
+        self.connections: dict[
+            HyperionWebsocketsRoom,
+            set[WebSocket],
+        ] = {}  # The Room and the set of users connected
+
+    async def connect_broadcaster(self):
+        await self.broadcaster.connect()
+
+    async def disconnect_broadcaster(self):
+        await self.broadcaster.disconnect()
+
+    async def add_connection_to_room(
+        self,
+        room_id: HyperionWebsocketsRoom,
+        ws_connection: WebSocket,
+    ) -> bool:
+        if room_id not in self.connections:
+            self.connections[room_id] = set({ws_connection})
+
+            subscribe_n_listen_task = asyncio.create_task(
+                self._subscribe_and_listen_to_channel(room_id=room_id),
+            )
+            wait_for_subscribe_task = asyncio.create_task(asyncio.sleep(1))
+
+            await asyncio.wait(
+                [subscribe_n_listen_task, wait_for_subscribe_task],
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+        else:
+            self.connections[room_id].add(ws_connection)
+
+        return True
+
+    async def _consume_events(self, message: str, room_id: HyperionWebsocketsRoom):
+        room_connections: set[WebSocket] = self.connections.get(room_id, set())
+        if len(room_connections) == 0:
+            return
+
+        for connection in room_connections:
+            await self._send_message_to_ws_connection(
+                message=f"Room {room_id} --> {message}",
+                ws_connection=connection,
+            )
+
+    async def _subscribe_and_listen_to_channel(self, room_id: str):
+        async with self.broadcaster.subscribe(channel=room_id) as subscriber:
+            async for event in subscriber:
+                message = MessageToRoomModel.model_validate_json(event.message)
+
+                await self._consume_events(
+                    message=message.message,
+                    room_id=message.room_id,
+                )
+
+    async def send_message_to_room(self, message: str, room_id: HyperionWebsocketsRoom):
+        room_connections: set[WebSocket] = self.connections.get(room_id, set())
+
+        if len(room_connections) == 0:
+            return
+
+        await self.broadcaster.publish(
+            channel=room_id,
+            message=MessageToRoomModel(
+                message=message,
+                room_id=room_id,
+            ).model_dump_json(),
+        )
+
+    async def _send_message_to_ws_connection(
+        self,
+        message: str,
+        ws_connection: WebSocket,
+    ) -> bool:
+        try:
+            await ws_connection.send_text(message)
+            return True
+        except RuntimeError:
+            for k in self.connections:
+                if ws_connection in self.connections[k]:
+                    self.connections[k].remove(ws_connection)
```

Comment:
Those conditions will cause the set to be resized, triggering an exception.
You can consider removing connections on the _consume_events function instead